### PR TITLE
Ensure rendering parity on DisplayAd, Comment, & Article

### LINKED ISSUE #18651: Refactor: DisplayAd should use rendering service
After #18649, `DisplayAd` should use the newly extracted rendering service.

Confirm:

* [ ] `DisplayAd` can include `LiquidTags`
* [ ] Same `AllowedAttributes`
* [ ] Same `AllowedTags`

### PR DESCRIPTION
**Pull request by lightalloy**
Ready for review, but I'll add more tests.

## What type of PR is this? (check all applicable)
- [x] Refactor

## Description
The `DisplayAd` rendering was similar to the `Article` and `Comment` ones, but a bit different. In this pr `DisplayAd` rendering becomes the same as `Article` and `Comments`, but preserves `DisplayAd` options.
I separated `BasicContentRenderer`(for `Comment` and `DisplayAd` - no frontmatter)  and `ContentRenderer` (with frontmatter). It's possible to have everything in one class but that would lead to have more parameters and conditions, and the code would be harder to understand.

We also have similar parsing for the Article's and Comment's previews, I decided that it's not a part of this pr, but we should implement it later (maybe when removing the feature flag?) I'm still not fan of the current implementation of `ContentRenderer`, but that is also not part of this pr.

## Related Tickets & Documents
- Related Issue #18904 
- Closes #18651

## QA Instructions, Screenshots, Recordings
- enable the `:consistent_rendering` FeatureFlag
- try to create `DisplayAd`s with liquid tags or regular display ads and check how they look 

### UI accessibility concerns?
Some of the liquid tags look weird, because they doesn't fit. Should we style them/limit liquid tag usage? Smth else?

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

### CODE REVIEW

File: app/services/markdown_processor/fixer/base.rb [hunk 1]
```diff
@@ -10,6 +10,9 @@ module Fixer
     class Base
       FRONT_MATTER_DETECTOR = /-{3}.*?-{3}/m
 
+      # Match @_username_ that is not preceded by backtick
+      USERNAME_WITH_UNDERSCORE_REGEXP = /(?<!`)@_\w+_/
```
Comment (lightalloy) [inline app/services/markdown_processor/fixer/base.rb:14]: rubocop

File: app/services/markdown_processor/parser.rb [hunk 2]
```diff
@@ -35,7 +37,8 @@ def finalize(link_attributes: {})
       code_tag_content = convert_code_tags_to_triple_backticks(@content)
       escaped_content = escape_liquid_tags_in_codeblock(code_tag_content)
       html = markdown.render(escaped_content)
-      sanitized_content = sanitize_rendered_markdown(html)
+      sanitized_content = ActionController::Base.helpers.sanitize html, @sanitize_options
```
Comment (lightalloy) [inline app/services/markdown_processor/parser.rb:40]: I used the method explicitly here, it's more conveniet to pass options, and it doesn't make sense to include the whole helper to use 1 method.

File: spec/services/markdown_processor/parser_spec.rb [hunk 3]
```diff
@@ -274,7 +274,7 @@ def generate_and_parse_markdown(raw_markdown)
   context "when provided with an @username" do
     context "when html has injected styles" do
       before do
-        create :user, username: "User1"
+        create(:user, username: "User1")
```
Comment (lightalloy) [inline spec/services/markdown_processor/parser_spec.rb:277]: rubocop

File: spec/services/basic_content_renderer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      "<h1>\n  <a name=\"hey\" href=\"#hey\" class=\"anchor\">\n  </a>\n  Hey\n</h1>\n\n<p>I'm a markdown</p>\n\n"
+    end
+
+    it "processes markdown" do
+      result = described_class.new(markdown, source: build(:comment), user: build(:user)).process
+      expect(result).to eq(expected_result)
+    end
+  end
+
+  context "when markdown has liquid tags that aren't allowed for source" do
```
Comment (lightalloy) [inline spec/services/basic_content_renderer_spec.rb:46]: Seems more like `MarkdownProcessor::Parser`s responsibility, but let's keep for "integration" testing (like for `ContentRenderer`)

**Review by jaw6 — COMMENTED** [review]
> I separated BasicContentRenderer(for Comment and DisplayAd - no frontmatter) and ContentRenderer (with frontmatter). It's possible to have everything in one class but that would lead to have more parameters and conditions, and the code would be harder to understand.

I'm probably missing something, but I'm having a hard time understanding `BasicContentRenderer`. I thought `ContentRenderer` can detect the presence of `frontmatter`, I'm probably not understanding something, but I'm not sure why there would need to be a parameter. The overarching goal of this milestone was to have our "content" all using a common rendering system, so I'm concerned that having two systems will undermine that goal. Can you elaborate on this part maybe?

> We also have similar parsing for the Article's and Comment's previews, I decided that it's not a part of this pr, but we should implement it later (maybe when removing the feature flag?) 

Great catch -- I opened a [new issue](https://github.com/forem/forem/issues/19099) in the milestone to capture this.

> I'm still not fan of the current implementation of ContentRenderer, but that is also not part of this pr.

Maybe we should chat about this some time, if there are issues with `ContentRenderer`, I think it would be good to hash them out sooner rather than later, given that it's still feature-flagged.

**Comment by lightalloy** [comment]
I think of this task first of all as "parse the DisplayAd the same way", and only then as "do everything in one class" (if do so).

It's possible to move in everything in one class, but there are several differences between parsing articles and other entities:
- parsing frontmatter: Does it make sense to parse frontmatter (`FrontMatterParser::Parser#call`) for display ads and comments if we know there will be no frontmatter (for performance and clarity)?
  + there were discussions about possibility of removing editor v1, so we may not need to parse frontmatter for article as well (though we'll still need to support old ones)
- calculating reading time (could be solved with a parameter to a `#process` method)
- for comments, we need to use a different fixer (so there is a constructor parameter in the `BasicContentRenderer`)
- `sanitize_options` for processing `DisplayAd` (a parameter for `BasicContentRenderer#process`)

So, if we put everything in one class, there would be more options and conditions (too much imo), but the main thing is that `Article`s stand out with their frontmatter.
I think that a bit of duplication is the lesser evil here.

As for the `ContentRenderer`, there are no huge issues, but it's not great that it sets state (front_matter and reading_time) and parses + returns processed html at the same time which is counterintuitive.
I would prefer using the result object, but I think it makes sense if `BasicContentRenderer#process` and `ContentRenderer#process` return the same type, so more changes needed if we choose to go that way. We can think of this as an opportunity for further refactoring.

Thanks for creating a separate issue for the previews!

**Comment by lightalloy** [comment]
Btw, I could work on the solution that would solve these 2 issues: having 2 renderers and modifying state + returning a separate result.
Maybe, as a next pr (or after the Billboards MVP)

**Comment by benhalpern** [comment]
@lightalloy this all seems well enough reasoned to me. I don't have strong opinions on content renderer vs 2, whatever the appropriate abstraction is. We're moving this away from the haphazard way it's currently done.

Another consideration we could circle back to later, or just handle now if it seems like it could help discovering the right abstractions, we _also_ do the same stuff with the `Page` model, in the simplest way of any of them:

```ruby
      parsed_markdown = MarkdownProcessor::Parser.new(body_markdown)
      self.processed_html = parsed_markdown.finalize
```

One last thing:

It looks to me like we'll lose this step in the proposed code for `display_ad`.

**_Code pasted from the bottom of `original_process_markdown` with the method below it also showing_**

```ruby
    self.processed_html = Html::Parser.new(html)
      .prefix_all_images(prefix_width, synchronous_detail_detection: true).html
  end

  def prefix_width
    placement_area.to_s == "post_comments" ? POST_WIDTH : SIDEBAR_WIDTH
  end
```

The inputs of `prefix_width` as calculated here and `synchronous_detail_detection: true` seem like they could be passed in as high-level config and then into `prefix_all_images` later where it is now called in the same place for all.

`synchronous_detail_detection` is only needed due to some funny inconsistencies with `Article`. For now, I think it would be best to try and pass these values in as optional config params for this new renderer so as to maintain current behavior. And future consolidation to how this all works could be a smaller task in the future, but would technically change behavior.

***

At the end of the day, we're looking to make the code cleaner than it is today, while ultimately allowing `liquid tags`, and other custom stuff we do to be generated everywhere as consistently as possible. And I think we're close to this.

**Review by maestromac — COMMENTED** [review]
I agree with @jaw6 . I think the presence of `BasicContentRenderer` can be a bit confusing and wonder if there's a better way to do it all under `ContentRenderer`.  And instead of extra parameter or options, you can try the more "a-la-carte" of constructing a method just for a particular use-case/class. Something along the line of how it's done in `Moderator::BanishUser`.

**Comment by lightalloy** [comment]
@benhalpern thank you for the `prefix_all_images` note, that's important! :raised_hands:

**Comment by lightalloy** [comment]
@maestromac we don't have a lot of cases in `Moderator::BanishUser` iself. Do you mean smth like `ManageActivityAndRoles#handle_user_status`? 
I think in the rendering case it's kind of an antipattern here to changed behaviour based on the class of the source.

I will be altering the solution anyway to fix the `prefix_all_images` (that Ben has mentioned), so I'll think of the alternate solution along with it.

**Comment by jaw6** [comment]
@lightalloy So, back in October, I had a lot of these same misgivings about how we might approach the task of making our rendering more consistent across different objects. We had a meeting about this and the strategy of building a common "rendering service" is that course we elected to pursue. 

I believe it would be a misstep, at this point, to introduce a second rendering service. It seems to me that many of the issues you bring up with `ContentRenderer` stem from the refactoring that came in with #18989. It seems to me that the refactoring there altered the design of this class in a way we should try to remedy if possible. I may be misunderstanding something, but I don't think the class design constraints should shift our overall approach in this area.

I think one rendering service _can_ work and, at this point, _should_ (continue to) be our strategy -- at least as long as Product's request is that these objects render in the same way.

File: app/models/display_ad.rb [hunk 5]
```diff
@@ -94,6 +94,26 @@ def generate_display_ad_name
...
+      extracted_process_markdown
+    else
+      original_process_markdown
+    end
+  end
+
+  def extracted_process_markdown
+    renderer = ContentRenderer.new(body_markdown || "", source: self)
+    sanitize_options = { tags: MarkdownProcessor::AllowedTags::DISPLAY_AD,
+                         attributes: MarkdownProcessor::AllowedAttributes::DISPLAY_AD }
```
Comment (jaw6) [inline app/models/display_ad.rb:109]: I could be wrong, but I think this should use the `Allowed<x>::FEED` constants.

File: app/services/content_renderer.rb [hunk 6]
```diff
@@ -1,37 +1,70 @@
+# renders markdown for Articles, DisplayAds, Comments
 class ContentRenderer
-  class_attribute :fixer, default: MarkdownProcessor::Fixer::FixAll
-  class_attribute :front_matter_parser, default: FrontMatterParser::Parser.new(:md)
+  Result = Struct.new(:front_matter, :reading_time, :processed_html, keyword_init: true)
```
Comment (jaw6) [inline app/services/content_renderer.rb:3]: I really dig this refactoring ✨

**Review by benhalpern — APPROVED** [review]
Tested locally and seems good to me 🎉

File: spec/services/content_renderer_spec.rb [hunk 7]
```diff
@@ -1,93 +1,88 @@
...
+    expect(MarkdownProcessor::Fixer::FixAll).to have_received(:call).with(markdown)
+  end
 
-      described_class.fixer = mock_fixer
-      described_class.front_matter_parser = mock_front_matter_parser
-      described_class.processor = mock_processor
+  describe "#process" do
+    let(:parser) { instance_double(MarkdownProcessor::Parser) }
+
+    context "with double parser" do
```
Comment (jaw6) [inline spec/services/content_renderer_spec.rb:16]: "double" parser here is a reference to `instance_double`, not having :two:x parsers?

**Review by jaw6 — APPROVED** [review]
Looks good to me! :+1:

**Pull request merged**